File: src/main/scala/simulations/BasicSimulation.scala
package simulations

import io.gatling.core.Predef._
import io.gatling.http.Predef._

import scala.util.Random
import scala.concurrent.duration._

class BasicSimulation extends Simulation {

  val httpProtocol = http.baseUrl("https://httpstat.us")

  def random() = Random.nextInt(5000)

  val scn = scenario("BasicSimulation")
    .exec(http("Should be ok").get("/200"))
    .exec(http("Gateway timeout").get("/504"))
    .exec(http("With random response time").get(s"/200?sleep=${random()}"))

  setUp(
    scn.inject(constantUsersPerSec(2) during(60 seconds))
  ).protocols(httpProtocol)

}

File: src/gatling/kotlin/simulations/BasicSimulation.kt
package simulations

import io.gatling.javaapi.core.CoreDsl.*
import io.gatling.javaapi.core.Simulation
import io.gatling.javaapi.http.HttpDsl.http
import io.gatling.javaapi.http.HttpDsl.status
import java.time.Duration.ofSeconds
import kotlin.time.Duration.Companion.seconds

class BasicSimulation : Simulation() {

    private val httpProtocol = http
        .baseUrl("https://testapi.jasonwatmore.com")
        .acceptHeader("application/json")
        .contentTypeHeader("application/json")

    private val getProducts = exec(
        http("GET all products")
            .get("/products")
            .check(status().`is`(200))
    )

    private val postProduct = exec(
        http("POST create product")
            .post("/products")
            .body(
                StringBody(
                    """
                    {
                        "name": "Gatling Test Product",
                        "price": "9.99"
                    }
                    """.trimIndent()
                )
            )
            .asJson()
            .check(status().`is`(200))
            .check(jsonPath("$.id").exists())
    )

    private val scn = scenario("GET and POST Real API")
        .exec(getProducts)
        .pause(1)
        .exec(postProduct)

    init {
        setUp(
            scn.injectOpen(constantUsersPerSec(3.0).during(ofSeconds(5)))
        ).protocols(httpProtocol)
    }
}
